src/database/repository/diesel/central_sync_cursor.rs

```
@@ -1,41 +1,36 @@
 
 use crate::database::{repository::RepositoryError, schema::CentralSyncCursorRow};
 
-use diesel::{
-    prelude::*,
-    r2d2::{ConnectionManager, Pool},
-};
+use diesel::prelude::*;
 
-pub struct CentralSyncCursorRepository {
-    pool: Pool<ConnectionManager<DBBackendConnection>>,
+pub struct CentralSyncCursorRepository<'a> {
+    connection: &'a StorageConnection,
 }
 
-impl CentralSyncCursorRepository {
-    pub fn new(pool: Pool<ConnectionManager<DBBackendConnection>>) -> CentralSyncCursorRepository {
-        CentralSyncCursorRepository { pool }
+impl<'a> CentralSyncCursorRepository<'a> {
+    pub fn new(connection: &'a StorageConnection) -> Self {
+        CentralSyncCursorRepository { connection }
     }
 
     pub async fn get_cursor(&self) -> Result<u32, RepositoryError> {
         use crate::database::schema::diesel_schema::central_sync_cursor::dsl::*;
-        let connection = get_connection(&self.pool)?;
-        let result: CentralSyncCursorRow = central_sync_cursor.first(&connection)?;
+        let result: CentralSyncCursorRow =
+            central_sync_cursor.first(&self.connection.connection)?;
         let cursor = result.id as u32;
         Ok(cursor)
     }
 
     pub async fn update_cursor(&self, cursor: u32) -> Result<(), RepositoryError> {
-        let connection = get_connection(&self.pool)?;
-        CentralSyncCursorRepository::update_cursor_tx(&connection, cursor)
-    }
-
-    pub fn update_cursor_tx(connection: &DBConnection, cursor: u32) -> Result<(), RepositoryError> {
         use crate::database::schema::diesel_schema::central_sync_cursor::dsl::*;
         let row = CentralSyncCursorRow { id: cursor as i32 };
-        diesel::delete(central_sync_cursor).execute(connection)?;
-        diesel::insert_into(central_sync_cursor)
-            .values(&row)
-            .execute(connection)?;
-        Ok(())
+        // note: if already in a transaction this creates a safepoint:
+        self.connection.connection.transaction(|| {
```

if an error happens, does it roll back to safepoint ?

---

yes but if the error is not handled in the parent transaction the whole transaction will get aborted